File: backend/utils/validators.py

"""
Validators - Complete validation system
"""

import os
import re
import json
import uuid
from datetime import datetime
from typing import Dict, Any, List, Optional, Tuple, Union
import logging

logger = logging.getLogger(__name__)

# ==================== CONSTANTS ====================

PATTERNS = {
    'email': r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$',
    'username': r'^[a-zA-Z0-9_.-]{3,30}$',
    'phone': r'^\+?[0-9]{10,15}$',
    'url': r'^https?://(?:[-\w.]|(?:%[\da-fA-F]{2}))+',
    'slug': r'^[a-z0-9]+(?:-[a-z0-9]+)*$',
    'alphanumeric': r'^[a-zA-Z0-9]+$',
    'alphanumeric_space': r'^[a-zA-Z0-9\s]+$',
    'numeric': r'^[0-9]+$',
    'decimal': r'^[0-9]+(\.[0-9]+)?$',
}

# ==================== SAFE IMPORTS ====================

try:
    import bleach
    HAS_BLEACH = True
except ImportError:
    HAS_BLEACH = False

try:
    import phonenumbers
    HAS_PHONENUMBERS = True
except ImportError:
    HAS_PHONENUMBERS = False

try:
    from email_validator import validate_email as validate_email_lib, EmailNotValidError
    HAS_EMAIL_VALIDATOR = True
except ImportError:
    HAS_EMAIL_VALIDATOR = False
    validate_email_lib = None
    EmailNotValidError = None

try:
    import imghdr
    HAS_IMGHDR = True
except ImportError:
    HAS_IMGHDR = False

# ==================== VALIDATION RESULT ====================

class ValidationResult:
    """Validation result container"""
    
    def __init__(self, valid: bool = True, errors: List[str] = None, 
                 data: Any = None, warnings: List[str] = None):
        self.valid = valid
        self.errors = errors or []
        self.data = data
        self.warnings = warnings or []
    
    def add_error(self, error: str):
        self.errors.append(error)
        self.valid = False
    
    def add_warning(self, warning: str):
        self.warnings.append(warning)
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            'valid': self.valid,
            'errors': self.errors,
            'warnings': self.warnings,
            'data': self.data
        }
    
    @classmethod
    def success(cls, data: Any = None, warnings: List[str] = None):
        return cls(valid=True, data=data, warnings=warnings)
    
    @classmethod
    def failure(cls, error: Union[str, List[str]], data: Any = None):
        errors = [error] if isinstance(error, str) else error
        return cls(valid=False, errors=errors, data=data)

# ==================== BASIC VALIDATORS ====================

def validate_required(value: Any, field_name: str = 'Field') -> ValidationResult:
    """Validate that a value is not empty"""
    if value is None or value == '' or value == [] or value == {}:
        return ValidationResult.failure(f"{field_name} is required")
    
    if isinstance(value, str) and not value.strip():
        return ValidationResult.failure(f"{field_name} is required")
    
    return ValidationResult.success(value)

def validate_string(value: Any, min_length: int = 0, max_length: int = None,
                    field_name: str = 'Field') -> ValidationResult:
    """Validate a string value"""
    result = validate_required(value, field_name)
    if not result.valid:
        return result
    
    if not isinstance(value, str):
        return ValidationResult.failure(f"{field_name} must be a string")
    
    if min_length > 0 and len(value) < min_length:
        return ValidationResult.failure(f"{field_name} must be at least {min_length} characters")
    
    if max_length and len(value) > max_length:
        return ValidationResult.failure(f"{field_name} must be at most {max_length} characters")
    
    return ValidationResult.success(value)

# ==================== EMAIL VALIDATOR ====================

def validate_email(email: str) -> ValidationResult:
    """Validate email address"""
    if not email:
        return ValidationResult.failure("Email is required")
    
    if not re.match(PATTERNS['email'], email):
        return ValidationResult.failure("Invalid email format")
    
    if HAS_EMAIL_VALIDATOR and validate_email_lib:
        try:
            validated = validate_email_lib(email, check_deliverability=False)
            return ValidationResult.success(validated.email)
        except EmailNotValidError as e:
            return ValidationResult.failure(f"Invalid email: {str(e)}")
        except Exception as e:
            logger.warning(f"Email validation error: {str(e)}")
    
    return ValidationResult.success(email)

# ==================== PHONE VALIDATOR ====================

def validate_phone(phone: str) -> ValidationResult:
    """Validate phone number"""
    if not phone:
        return ValidationResult.failure("Phone number is required")
    
    if HAS_PHONENUMBERS:
        try:
            phone_clean = re.sub(r'[\s\-()]', '', phone)
            parsed = phonenumbers.parse(phone_clean, None)
            if not phonenumbers.is_valid_number(parsed):
                return ValidationResult.failure("Invalid phone number")
            formatted = phonenumbers.format_number(
                parsed, 
                phonenumbers.PhoneNumberFormat.INTERNATIONAL
            )
            return ValidationResult.success(formatted)
        except:
            pass
    
    # Fallback to regex
    if re.match(PATTERNS['phone'], phone):
        return ValidationResult.success(phone)
    return ValidationResult.failure("Invalid phone number")

# ==================== PASSWORD VALIDATOR ====================

def validate_password(password: str, min_length: int = 8) -> ValidationResult:
    """Validate password strength"""
    if not password:
        return ValidationResult.failure("Password is required")
    
    errors = []
    warnings = []
    strength_score = 0
    
    if len(password) < min_length:
        errors.append(f"Password must be at least {min_length} characters long")
    elif len(password) >= 12:
        strength_score += 1
    
    if not re.search(r'[A-Z]', password):
        errors.append("Password must contain at least one uppercase letter")
    else:
        strength_score += 1
    
    if not re.search(r'[a-z]', password):
        errors.append("Password must contain at least one lowercase letter")
    else:
        strength_score += 1
    
    if not re.search(r'[0-9]', password):
        errors.append("Password must contain at least one number")
    else:
        strength_score += 1
    
    if not re.search(r'[!@#$%^&*(),.?":{}|<>]', password):
        warnings.append("Password should contain at least one special character")
    else:
        strength_score += 1
    
    if strength_score >= 5:
        strength = 'strong'
    elif strength_score >= 3:
        strength = 'medium'
    else:
        strength = 'weak'
    
    if errors:
        return ValidationResult.failure(errors)
    
    result = ValidationResult.success(password)
    result.data = {'strength': strength, 'score': strength_score}
    result.warnings = warnings
    return result

# ==================== USERNAME VALIDATOR ====================

def validate_username(username: str) -> ValidationResult:
    """Validate username"""
    if not username:
        return ValidationResult.failure("Username is required")
    
    if len(username) < 3:
        return ValidationResult.failure("Username must be at least 3 characters")
    
    if len(username) > 30:
        return ValidationResult.failure("Username must be at most 30 characters")
    
    if not re.match(PATTERNS['username'], username):
        return ValidationResult.failure("Username can only contain letters, numbers, dots, underscores, and hyphens")
    
    return ValidationResult.success(username)

# ==================== URL VALIDATOR ====================

def validate_url(url: str) -> ValidationResult:
    """Validate URL"""
    if not url:
        return ValidationResult.failure("URL is required")
    
    if not re.match(PATTERNS['url'], url):
        return ValidationResult.failure("Invalid URL format")
    
    return ValidationResult.success(url)

# ==================== DATE VALIDATORS ====================

def validate_date(date_str: str, format: str = '%Y-%m-%d') -> ValidationResult:
    """Validate date string"""
    if not date_str:
        return ValidationResult.failure("Date is required")
    
    try:
        date_obj = datetime.strptime(date_str, format)
        return ValidationResult.success(date_obj)
    except ValueError:
        return ValidationResult.failure(f"Invalid date format. Expected {format}")

def validate_datetime(dt_str: str) -> ValidationResult:
    """Validate datetime string (ISO format)"""
    if not dt_str:
        return ValidationResult.failure("Datetime is required")
    
    try:
        dt_obj = datetime.fromisoformat(dt_str.replace('Z', '+00:00'))
        return ValidationResult.success(dt_obj)
    except ValueError:
        return ValidationResult.failure("Invalid datetime format")

def validate_future_date(date_obj: datetime, allow_today: bool = False) -> ValidationResult:
    """Validate that date is in the future"""
    if not date_obj:
        return ValidationResult.failure("Date is required")
    
    now = datetime.utcnow()
    if allow_today:
        if date_obj < now.replace(hour=0, minute=0, second=0, microsecond=0):
            return ValidationResult.failure("Date must be today or in the future")
    else:
        if date_obj <= now:
            return ValidationResult.failure("Date must be in the future")
    
    return ValidationResult.success(date_obj)

# ==================== NUMBER VALIDATORS ====================

def validate_number(value: Any, min_val: float = None, max_val: float = None,
                    field_name: str = 'Field') -> ValidationResult:
    """Validate numeric value"""
    if value is None:
        return ValidationResult.failure(f"{field_name} is required")
    
    try:
        num = float(value)
    except (ValueError, TypeError):
        return ValidationResult.failure(f"{field_name} must be a number")
    
    if min_val is not None and num < min_val:
        return ValidationResult.failure(f"{field_name} must be at least {min_val}")
    
    if max_val is not None and num > max_val:
        return ValidationResult.failure(f"{field_name} must be at most {max_val}")
    
    return ValidationResult.success(num)

def validate_integer(value: Any, min_val: int = None, max_val: int = None,
                     field_name: str = 'Field') -> ValidationResult:
    """Validate integer value"""
    if value is None:
        return ValidationResult.failure(f"{field_name} is required")
    
    try:
        num = int(value)
    except (ValueError, TypeError):
        return ValidationResult.failure(f"{field_name} must be an integer")
    
    if min_val is not None and num < min_val:
        return ValidationResult.failure(f"{field_name} must be at least {min_val}")
    
    if max_val is not None and num > max_val:
        return ValidationResult.failure(f"{field_name} must be at most {max_val}")
    
    return ValidationResult.success(num)

# ==================== RATING VALIDATOR ====================

def validate_rating(rating: int) -> ValidationResult:
    """Validate rating (1-5)"""
    if rating is None:
        return ValidationResult.failure("Rating is required")
    
    try:
        rating = int(rating)
    except (ValueError, TypeError):
        return ValidationResult.failure("Rating must be a number")
    
    if rating < 1 or rating > 5:
        return ValidationResult.failure("Rating must be between 1 and 5")
    
    return ValidationResult.success(rating)

# ==================== REVIEW TEXT VALIDATOR ====================

def validate_review_text(text: str, min_length: int = 0, max_length: int = 2000) -> ValidationResult:
    """
    Validate review text
    
    Args:
        text: Review text to validate
        min_length: Minimum length (default: 0)
        max_length: Maximum length (default: 2000)
    
    Returns:
        ValidationResult
    """
    if text is None:
        return ValidationResult.success('')
    
    if not isinstance(text, str):
        return ValidationResult.failure("Review text must be a string")
    
    # Check length
    if len(text) < min_length:
        return ValidationResult.failure(f"Review text must be at least {min_length} characters")
    
    if len(text) > max_length:
        return ValidationResult.failure(f"Review text must be at most {max_length} characters")
    
    # Sanitize text (remove excessive whitespace)
    text = ' '.join(text.split())
    
    return ValidationResult.success(text)

# ==================== SEAT AND PRICE VALIDATORS ====================

def validate_seats(seats: int, max_seats: int = 10000) -> ValidationResult:
    """Validate seat count"""
    if seats is None:
        return ValidationResult.failure("Seats count is required")
    
    try:
        seats = int(seats)
    except (ValueError, TypeError):
        return ValidationResult.failure("Seats must be a number")
    
    if seats <= 0:
        return ValidationResult.failure("Seats must be greater than 0")
    
    if seats > max_seats:
        return ValidationResult.failure(f"Seats cannot exceed {max_seats}")
    
    return ValidationResult.success(seats)

def validate_price(price: float) -> ValidationResult:
    """Validate price"""
    if price is None:
        return ValidationResult.failure("Price is required")
    
    try:
        price = float(price)
    except (ValueError, TypeError):
        return ValidationResult.failure("Price must be a number")
    
    if price < 0:
        return ValidationResult.failure("Price cannot be negative")
    
    price = round(price, 2)
    
    return ValidationResult.success(price)

# ==================== EVENT DATE VALIDATOR ====================

def validate_event_date(event_date: datetime, registration_deadline: datetime = None) -> ValidationResult:
    """Validate event date with registration deadline"""
    if not event_date:
        return ValidationResult.failure("Event date is required")
    
    result = validate_future_date(event_date, allow_today=True)
    if not result.valid:
        return result
    
    if registration_deadline:
        if registration_deadline >= event_date:
            return ValidationResult.failure("Registration deadline must be before event date")
        
        if registration_deadline <= datetime.utcnow():
            return ValidationResult.failure("Registration deadline has already passed")
    
    return ValidationResult.success(event_date)

# ==================== BOOKING QUANTITY VALIDATOR ====================

def validate_booking_quantity(quantity: int, available_seats: int, max_per_booking: int = 10) -> ValidationResult:
    """Validate booking quantity"""
    if quantity is None:
        return ValidationResult.failure("Quantity is required")
    
    try:
        quantity = int(quantity)
    except (ValueError, TypeError):
        return ValidationResult.failure("Quantity must be a number")
    
    if quantity <= 0:
        return ValidationResult.failure("Quantity must be at least 1")
    
    if quantity > max_per_booking:
        return ValidationResult.failure(f"Maximum {max_per_booking} tickets per booking")
    
    if quantity > available_seats:
        return ValidationResult.failure(f"Only {available_seats} seats available")
    
    return ValidationResult.success(quantity)

# ==================== SANITIZERS ====================

def sanitize_html(html: str) -> str:
    """Sanitize HTML content"""
    if not html:
        return ''
    
    if HAS_BLEACH:
        allowed_tags = [
            'p', 'br', 'strong', 'em', 'u', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6',
            'ul', 'ol', 'li', 'a', 'img', 'blockquote', 'code', 'pre', 'span', 'div'
        ]
        allowed_attributes = {
            'a': ['href', 'target', 'rel'],
            'img': ['src', 'alt', 'width', 'height'],
        }
        return bleach.clean(html, tags=allowed_tags, attributes=allowed_attributes, strip=True)
    
    return html

def sanitize_input(text: str, max_length: int = 1000) -> str:
    """Sanitize input text"""
    if not text:
        return ''
    
    text = re.sub(r'[<>]', '', text)
    
    if len(text) > max_length:
        text = text[:max_length]
    
    return text.strip()

def sanitize_filename(filename: str) -> str:
    """Sanitize filename"""
    if not filename:
        return ''
    
    filename = os.path.basename(filename)
    filename = re.sub(r'[<>:"/\\|?*]', '', filename)
    
    if len(filename) > 255:
        name, ext = os.path.splitext(filename)
        filename = name[:250] + ext
    
    return filename

# ==================== COMPOSITE VALIDATORS ====================

class ValidatorChain:
    """Chain multiple validators together"""
    
    def __init__(self, validators: List[callable] = None):
        self.validators = validators or []
    
    def add(self, validator: callable) -> 'ValidatorChain':
        self.validators.append(validator)
        return self
    
    def validate(self, value: Any) -> ValidationResult:
        current_value = value
        all_errors = []
        all_warnings = []
        
        for validator in self.validators:
            result = validator(current_value)
            if not result.valid:
                all_errors.extend(result.errors)
            if result.warnings:
                all_warnings.extend(result.warnings)
            if result.data is not None and result.data != current_value:
                current_value = result.data
        
        if all_errors:
            return ValidationResult.failure(all_errors, current_value)
        
        return ValidationResult.success(current_value, all_warnings)

# ==================== VALIDATOR FACTORY ====================

class ValidatorFactory:
    """Factory for creating validator chains"""
    
    @staticmethod
    def create_user_validator() -> ValidatorChain:
        return ValidatorChain([
            lambda data: validate_required(data, 'User data'),
            lambda data: validate_dict(data, ['username', 'email', 'password']),
            lambda data: validate_username(data.get('username')),
            lambda data: validate_email(data.get('email')),
            lambda data: validate_password(data.get('password')),
        ])
    
    @staticmethod
    def create_event_validator() -> ValidatorChain:
        return ValidatorChain([
            lambda data: validate_required(data, 'Event data'),
            lambda data: validate_dict(data, ['title', 'description', 'category_id', 'date']),
            lambda data: validate_string(data.get('title'), 1, 200, 'Title'),
            lambda data: validate_string(data.get('description'), 1, 5000, 'Description'),
            lambda data: validate_integer(data.get('category_id'), 1, 'Category ID'),
            lambda data: validate_date(data.get('date')),
        ])
    
    @staticmethod
    def create_booking_validator() -> ValidatorChain:
        return ValidatorChain([
            lambda data: validate_required(data, 'Booking data'),
            lambda data: validate_dict(data, ['event_id', 'quantity']),
            lambda data: validate_integer(data.get('event_id'), 1, 'Event ID'),
            lambda data: validate_integer(data.get('quantity'), 1, 10, 'Quantity'),
        ])

# ==================== DICT VALIDATOR ====================

def validate_dict(data: Dict, required_fields: List[str] = None) -> ValidationResult:
    """Validate dictionary"""
    if data is None:
        return ValidationResult.failure("Data is required")
    
    if not isinstance(data, dict):
        return ValidationResult.failure("Data must be a dictionary")
    
    errors = []
    
    if required_fields:
        for field in required_fields:
            if field not in data or data[field] is None or data[field] == '':
                errors.append(f"Required field '{field}' is missing")
    
    if errors:
        return ValidationResult.failure(errors)
    
    return ValidationResult.success(data)

# ==================== INITIALIZATION ====================

def init_validators(app):
    """Initialize validators with Flask app"""
    app.config['VALIDATORS'] = {
        'email': validate_email,
        'phone': validate_phone,
        'password': validate_password,
        'username': validate_username,
        'url': validate_url,
        'date': validate_date,
        'datetime': validate_datetime,
        'rating': validate_rating,
        'review_text': validate_review_text,
        'seats': validate_seats,
        'price': validate_price,
    }
    
    logger.info("Validators initialized successfully")
    return app

# ==================== EXPORTS ====================

__all__ = [
    'ValidationResult',
    'validate_required',
    'validate_string',
    'validate_email',
    'validate_phone',
    'validate_password',
    'validate_username',
    'validate_url',
    'validate_date',
    'validate_datetime',
    'validate_future_date',
    'validate_number',
    'validate_integer',
    'validate_rating',
    'validate_review_text',
    'validate_seats',
    'validate_price',
    'validate_event_date',
    'validate_booking_quantity',
    'sanitize_html',
    'sanitize_input',
    'sanitize_filename',
    'ValidatorChain',
    'ValidatorFactory',
    'validate_dict',
    'init_validators',
]

File: frontend/src/utils/validators.js
/**
 * Validation utilities for forms and data
 */

// ==================== EMAIL VALIDATION ====================

export const isValidEmail = (email) => {
  const emailRegex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/
  return emailRegex.test(email)
}

// ==================== PASSWORD VALIDATION ====================

export const validatePassword = (password) => {
  const errors = []
  const strength = {
    score: 0,
    level: 'weak',
    message: ''
  }

  if (!password || password.length === 0) {
    errors.push('Password is required')
    return { valid: false, errors, strength }
  }

  // Length check
  if (password.length < 8) {
    errors.push('Password must be at least 8 characters')
  } else {
    strength.score += 1
  }

  // Uppercase check
  if (!/[A-Z]/.test(password)) {
    errors.push('Password must contain at least one uppercase letter')
  } else {
    strength.score += 1
  }

  // Lowercase check
  if (!/[a-z]/.test(password)) {
    errors.push('Password must contain at least one lowercase letter')
  } else {
    strength.score += 1
  }

  // Number check
  if (!/[0-9]/.test(password)) {
    errors.push('Password must contain at least one number')
  } else {
    strength.score += 1
  }

  // Special character check
  if (!/[!@#$%^&*(),.?":{}|<>]/.test(password)) {
    errors.push('Password must contain at least one special character')
  } else {
    strength.score += 1
  }

  // Determine strength level
  if (strength.score >= 5) {
    strength.level = 'excellent'
    strength.message = 'Strong password'
  } else if (strength.score >= 4) {
    strength.level = 'good'
    strength.message = 'Good password'
  } else if (strength.score >= 3) {
    strength.level = 'fair'
    strength.message = 'Fair password'
  } else if (strength.score >= 1) {
    strength.level = 'weak'
    strength.message = 'Weak password'
  } else {
    strength.level = 'very-weak'
    strength.message = 'Very weak password'
  }

  return {
    valid: errors.length === 0,
    errors,
    strength
  }
}

// ==================== USERNAME VALIDATION ====================

export const isValidUsername = (username) => {
  if (!username || username.length === 0) {
    return { valid: false, error: 'Username is required' }
  }

  if (username.length < 3) {
    return { valid: false, error: 'Username must be at least 3 characters' }
  }

  if (username.length > 30) {
    return { valid: false, error: 'Username must be at most 30 characters' }
  }

  if (!/^[a-zA-Z0-9_.-]+$/.test(username)) {
    return { valid: false, error: 'Username can only contain letters, numbers, dots, underscores, and hyphens' }
  }

  return { valid: true }
}

// ==================== PHONE VALIDATION ====================

export const isValidPhone = (phone) => {
  if (!phone) return { valid: true } // Phone is optional

  // Remove spaces, dashes, parentheses
  const cleaned = phone.replace(/[\s\-()]/g, '')
  
  if (!/^\+?[0-9]{10,15}$/.test(cleaned)) {
    return { valid: false, error: 'Invalid phone number format' }
  }

  return { valid: true }
}

// ==================== URL VALIDATION ====================

export const isValidUrl = (url) => {
  if (!url) return { valid: true }

  try {
    const parsed = new URL(url)
    return { valid: parsed.protocol === 'http:' || parsed.protocol === 'https:' }
  } catch {
    return { valid: false, error: 'Invalid URL format' }
  }
}

// ==================== DATE VALIDATION ====================

export const isValidDate = (dateString) => {
  if (!dateString) return { valid: false, error: 'Date is required' }

  const date = new Date(dateString)
  if (isNaN(date.getTime())) {
    return { valid: false, error: 'Invalid date format' }
  }

  return { valid: true, date }
}

export const isFutureDate = (dateString) => {
  const result = isValidDate(dateString)
  if (!result.valid) return result

  const now = new Date()
  now.setHours(0, 0, 0, 0)
  
  if (result.date < now) {
    return { valid: false, error: 'Date must be in the future' }
  }

  return { valid: true, date: result.date }
}

export const isPastDate = (dateString) => {
  const result = isValidDate(dateString)
  if (!result.valid) return result

  const now = new Date()
  now.setHours(0, 0, 0, 0)
  
  if (result.date > now) {
    return { valid: false, error: 'Date must be in the past' }
  }

  return { valid: true, date: result.date }
}

// ==================== NUMBER VALIDATION ====================

export const isValidNumber = (value, options = {}) => {
  const { min, max, required = true, fieldName = 'Value' } = options

  if (required && (value === undefined || value === null || value === '')) {
    return { valid: false, error: `${fieldName} is required` }
  }

  const num = Number(value)
  if (isNaN(num)) {
    return { valid: false, error: `${fieldName} must be a number` }
  }

  if (min !== undefined && num < min) {
    return { valid: false, error: `${fieldName} must be at least ${min}` }
  }

  if (max !== undefined && num > max) {
    return { valid: false, error: `${fieldName} must be at most ${max}` }
  }

  return { valid: true, value: num }
}

// ==================== RATING VALIDATION ====================

export const isValidRating = (rating) => {
  const result = isValidNumber(rating, { min: 1, max: 5, fieldName: 'Rating' })
  if (!result.valid) return result
  return { valid: true, value: Math.round(result.value) }
}

// ==================== SEAT VALIDATION ====================

export const isValidSeats = (seats, maxSeats = 10000) => {
  const result = isValidNumber(seats, { min: 1, max: maxSeats, fieldName: 'Seats' })
  if (!result.valid) return result
  return { valid: true, value: Math.floor(result.value) }
}

// ==================== PRICE VALIDATION ====================

export const isValidPrice = (price) => {
  const result = isValidNumber(price, { min: 0, fieldName: 'Price' })
  if (!result.valid) return result
  return { valid: true, value: Math.round(result.value * 100) / 100 }
}

// ==================== ZIP CODE VALIDATION ====================

export const isValidZipCode = (zipCode) => {
  if (!zipCode) return { valid: true }
  const cleaned = zipCode.replace(/\s/g, '')
  if (!/^[0-9]{5}(-[0-9]{4})?$/.test(cleaned)) {
    return { valid: false, error: 'Invalid zip code format' }
  }
  return { valid: true }
}

// ==================== NAME VALIDATION ====================

export const isValidName = (name, fieldName = 'Name') => {
  if (!name || name.trim().length === 0) {
    return { valid: false, error: `${fieldName} is required` }
  }

  if (name.trim().length < 2) {
    return { valid: false, error: `${fieldName} must be at least 2 characters` }
  }

  if (name.trim().length > 100) {
    return { valid: false, error: `${fieldName} must be at most 100 characters` }
  }

  if (!/^[a-zA-Z\s\-']+$/.test(name.trim())) {
    return { valid: false, error: `${fieldName} contains invalid characters` }
  }

  return { valid: true, value: name.trim() }
}

// ==================== DESCRIPTION VALIDATION ====================

export const isValidDescription = (description, options = {}) => {
  const { min = 10, max = 5000, required = true, fieldName = 'Description' } = options

  if (required && (!description || description.trim().length === 0)) {
    return { valid: false, error: `${fieldName} is required` }
  }

  if (description && description.trim().length < min) {
    return { valid: false, error: `${fieldName} must be at least ${min} characters` }
  }

  if (description && description.trim().length > max) {
    return { valid: false, error: `${fieldName} must be at most ${max} characters` }
  }

  return { valid: true, value: description ? description.trim() : '' }
}

// ==================== FORM VALIDATION ====================

export const validateForm = (data, rules) => {
  const errors = {}
  let isValid = true

  for (const [field, rule] of Object.entries(rules)) {
    const value = data[field]
    const result = rule(value)
    if (!result.valid) {
      errors[field] = result.error
      isValid = false
    }
  }

  return { isValid, errors }
}

// ==================== REGISTRATION VALIDATION ====================

export const validateRegistration = (data) => {
  const errors = {}

  // Username
  const usernameResult = isValidUsername(data.username)
  if (!usernameResult.valid) {
    errors.username = usernameResult.error
  }

  // Email
  if (!data.email) {
    errors.email = 'Email is required'
  } else if (!isValidEmail(data.email)) {
    errors.email = 'Please enter a valid email address'
  }

  // Password
  const passwordResult = validatePassword(data.password)
  if (!passwordResult.valid) {
    errors.password = passwordResult.errors[0]
  }

  // Confirm Password
  if (data.password !== data.confirmPassword) {
    errors.confirmPassword = 'Passwords do not match'
  }

  // Phone (optional)
  if (data.phone) {
    const phoneResult = isValidPhone(data.phone)
    if (!phoneResult.valid) {
      errors.phone = phoneResult.error
    }
  }

  // Terms
  if (!data.agreeTerms) {
    errors.agreeTerms = 'You must agree to the terms and conditions'
  }

  return {
    isValid: Object.keys(errors).length === 0,
    errors
  }
}

// ==================== EVENT VALIDATION ====================

export const validateEvent = (data) => {
  const errors = {}

  // Title
  if (!data.title || data.title.trim().length === 0) {
    errors.title = 'Event title is required'
  } else if (data.title.trim().length < 3) {
    errors.title = 'Event title must be at least 3 characters'
  } else if (data.title.trim().length > 200) {
    errors.title = 'Event title must be at most 200 characters'
  }

  // Description
  if (!data.description || data.description.trim().length === 0) {
    errors.description = 'Event description is required'
  } else if (data.description.trim().length < 20) {
    errors.description = 'Event description must be at least 20 characters'
  } else if (data.description.trim().length > 5000) {
    errors.description = 'Event description must be at most 5000 characters'
  }

  // Category
  if (!data.category_id) {
    errors.category_id = 'Please select a category'
  }

  // Venue
  if (!data.venue || data.venue.trim().length === 0) {
    errors.venue = 'Venue is required'
  }

  // City
  if (!data.city || data.city.trim().length === 0) {
    errors.city = 'City is required'
  }

  // Date
  if (!data.date) {
    errors.date = 'Event date is required'
  } else {
    const dateResult = isFutureDate(data.date)
    if (!dateResult.valid) {
      errors.date = dateResult.error
    }
  }

  // Total Seats
  if (!data.total_seats) {
    errors.total_seats = 'Total seats is required'
  } else {
    const seatsResult = isValidSeats(data.total_seats)
    if (!seatsResult.valid) {
      errors.total_seats = seatsResult.error
    }
  }

  // Price
  if (data.price !== undefined && data.price !== null) {
    const priceResult = isValidPrice(data.price)
    if (!priceResult.valid) {
      errors.price = priceResult.error
    }
  }

  return {
    isValid: Object.keys(errors).length === 0,
    errors
  }
}

// ==================== BOOKING VALIDATION ====================

export const validateBooking = (data) => {
  const errors = {}

  if (!data.event_id) {
    errors.event_id = 'Event ID is required'
  }

  if (!data.quantity) {
    errors.quantity = 'Quantity is required'
  } else {
    const quantityResult = isValidNumber(data.quantity, { min: 1, max: 10, fieldName: 'Quantity' })
    if (!quantityResult.valid) {
      errors.quantity = quantityResult.error
    }
  }

  return {
    isValid: Object.keys(errors).length === 0,
    errors
  }
}

// ==================== REVIEW VALIDATION ====================

export const validateReview = (data) => {
  const errors = {}

  if (!data.rating) {
    errors.rating = 'Rating is required'
  } else {
    const ratingResult = isValidRating(data.rating)
    if (!ratingResult.valid) {
      errors.rating = ratingResult.error
    }
  }

  if (data.review && data.review.trim().length > 2000) {
    errors.review = 'Review must be at most 2000 characters'
  }

  return {
    isValid: Object.keys(errors).length === 0,
    errors
  }
}

// ==================== EXPORT ALL ====================

export default {
  isValidEmail,
  validatePassword,
  isValidUsername,
  isValidPhone,
  isValidUrl,
  isValidDate,
  isFutureDate,
  isPastDate,
  isValidNumber,
  isValidRating,
  isValidSeats,
  isValidPrice,
  isValidZipCode,
  isValidName,
  isValidDescription,
  validateForm,
  validateRegistration,
  validateEvent,
  validateBooking,
  validateReview
}